drake_ros_systems/src/scene_markers_system.cpp

```diff
@@ -0,0 +1,352 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <drake/common/drake_copyable.h>
+#include <drake/common/eigen_types.h>
+#include <drake/geometry/geometry_properties.h>
+#include <drake/geometry/geometry_roles.h>
+#include <drake/geometry/query_object.h>
+#include <drake/geometry/scene_graph.h>
+#include <drake/geometry/scene_graph_inspector.h>
+#include <drake/geometry/shape_specification.h>
+#include <drake/math/rigid_transform.h>
+#include <drake/systems/framework/leaf_system.h>
+
+#include <rclcpp/duration.hpp>
+#include <rclcpp/time.hpp>
+
+#include <builtin_interfaces/msg/time.hpp>
+#include <visualization_msgs/msg/marker.hpp>
+#include <visualization_msgs/msg/marker_array.hpp>
+
+#include "drake_ros_systems/scene_markers_system.hpp"
+#include "drake_ros_systems/utilities/type_conversion.hpp"
+
+
+namespace drake_ros_systems
+{
+
+namespace
+{
+
+class SceneGeometryToMarkers : public drake::geometry::ShapeReifier
+{
+public:
+  DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(SceneGeometryToMarkers)
+
+  SceneGeometryToMarkers(
+    const drake::geometry::Role & role,
+    const drake::geometry::Rgba & default_color)
+  : role_(role), default_color_(default_color)
+  {
+  }
+
+  ~SceneGeometryToMarkers() override = default;
+
+  void Populate(
+    const drake::geometry::SceneGraphInspector<double> & inspector,
+    const drake::geometry::GeometryId & geometry_id,
+    visualization_msgs::msg::MarkerArray * marker_array)
+  {
+    DRAKE_ASSERT(nullptr != marker_array);
+
+    marker_array_ = marker_array;
+
+    prototype_marker_.header.frame_id =
+      inspector.GetName(inspector.GetFrameId(geometry_id));
+    prototype_marker_.ns = inspector.GetOwningSourceName(geometry_id) +
+      "::" + inspector.GetName(geometry_id);
```

Comment:
Isn't this going to generate a separate namespace for each marker? That's not very usable if people are trying to visualize a larger system and want to select/deselect chunks of it.

Reply:
It will, yes. I considered using source names as marker namespaces and geometry IDs as marker IDs, but:

a. Geometry IDs are 64 bit integers whereas marker IDs are 32 bit integers. This is relatively minor, as I'd expect things to fall apart way before one can attempt to visualize 2^31 + 1 markers.
b. To render a single, capsule-shaped geometry, three (3) markers are necessary. And with that, the correlation between marker ID and geometry ID is lost.

I guess we could allocate 29 bits for the ID, 2 bits to distinguish capsule geometries' constitutive markers, and throw if bit truncation of the geometry ID would lose information. Then we could afford putting all markers for all geometries from a given source in its namespace. WDYT?

Reply:
We've made the choice that marker namespaces are useful, but ids within that namespace are entirely opaque and don't have meaning - basically the approach of
```
marker.id = static_cast<int>(marker_array.markers.size() + 1);
marker_array.markers.push_back(marker);
```
In the general "visualize a whole SceneGraph" case, we use namespace = source name, and in other cases we break it down by body/frame names to allow markers corresponding to specific bodies to be selected. See `SceneGraphCollisionChecker::MakeContextRvizRepresentationOfConfigurationCollisions` for the most extreme case of this.

Reply:
That's reasonable approach, but it won't work as-is if the geometries are being added and removed dynamically from the `SceneGraph`. Neither will this patch, though it's easy to adapt it (i.e. to track which geometry IDs have been published already).

Reply:
That's an interesting approach. Have you considered submitting that patch to upstream Rviz?

Reply:
https://github.com/ros2/rviz/pull/685 was merged and released into Rolling. Feature's there.